service/frontend/task_reachability.go

```diff
@@ -0,0 +1,368 @@
+		setIdx := -1
+		buildIdIdx := -1
+		if len(request.versionSets) > 0 {
+			for sidx, set := range request.versionSets {
+				for bidx, id := range set.BuildIds {
+					if request.buildId == id {
+						setIdx = sidx
+						buildIdIdx = bidx
+						break
+					}
+				}
+			}
+		}
+
+		if setIdx == -1 {
+			// build id not in set - unreachable
+			return &taskQueueReachability, nil
+		}
+		set := request.versionSets[setIdx]
+		isDefaultInSet := buildIdIdx == len(set.BuildIds)-1
+
+		if !isDefaultInSet {
+			// unreachable
+			return &taskQueueReachability, nil
+		}
+
+		isDefaultInQueue = setIdx == len(request.versionSets)-1
+		escapedBuildIds := make([]string, len(set.GetBuildIds()))
+		for i, buildId := range set.GetBuildIds() {
+			escapedBuildIds[i] = fmt.Sprintf("%q", buildId)
+		}
+		buildIdsFilter = fmt.Sprintf("BuildIds IN (%s)", strings.Join(escapedBuildIds, ","))
+		// TODO: To properly answer if there are open workflows that may be routed to a build id, we'll need to see if
+		// any of them have not *yet* been assigned a build id, e.g. they've been started but have not had any completed
+		// workflow tasks.
+		// Since this is an API used to check which build ids can be retired, we'll take the stricter approach and
+		// assume that `BuildIds = NULL` does not mean old unversioned workflows.
+		if isDefaultInQueue {
+			buildIdsFilter = fmt.Sprintf("(%s OR BuildIds IS NULL)", buildIdsFilter)
+		}
+	}
+
+	if isDefaultInQueue {
+		taskQueueReachability.Reachability = append(
+			taskQueueReachability.Reachability,
+			enumspb.TASK_REACHABILITY_NEW_WORKFLOWS,
+		)
+	}
+
+	reachability, err := wh.queryVisibilityForExisitingWorkflowsReachability(ctx, request.namespace, request.taskQueue, buildIdsFilter)
+	if err != nil {
+		return nil, err
+	}
+	taskQueueReachability.Reachability = append(taskQueueReachability.Reachability, reachability...)
+	return &taskQueueReachability, nil
+}
+
+func (wh *WorkflowHandler) queryVisibilityForExisitingWorkflowsReachability(
+	ctx context.Context,
+	ns *namespace.Namespace,
+	taskQueue,
+	buildIdsFilter string,
+) ([]enumspb.TaskReachability, error) {
+	type reachabilityQuery struct {
+		str              string
+		taskReachability enumspb.TaskReachability
+	}
+
+	queries := []reachabilityQuery{
+		{
+			str:              fmt.Sprintf("TaskQueue = %q AND %s AND ExecutionStatus = \"Running\"", taskQueue, buildIdsFilter),
+			taskReachability: enumspb.TASK_REACHABILITY_OPEN_WORKFLOWS,
+		},
+		{
+			str:              fmt.Sprintf("TaskQueue = %q AND %s AND ExecutionStatus != \"Running\"", taskQueue, buildIdsFilter),
+			taskReachability: enumspb.TASK_REACHABILITY_CLOSED_WORKFLOWS,
+		},
+	}
+	reachabilityResponses, err := raceMap(queries, func(query reachabilityQuery) (enumspb.TaskReachability, error) {
+		req := manager.CountWorkflowExecutionsRequest{
+			NamespaceID: ns.ID(),
+			Namespace:   ns.Name(),
+			Query:       query.str,
+		}
+
+		// TODO: is count more efficient than select with page size of 1?
```

Comment:
I think select with limit of one is more efficient, not sure... @rodrigozhou or @alexshtin could verify.

Reply:
very possible the count is more efficient, esp. if it can be approximate. databases often keep summary stats to aid the query planner and the count might be able to be answered from those without touching any data at all